### PR TITLE
replaces findIndices with elemIndices

### LINKED ISSUE #14: `lineSplit` inefficiency using ByteString `findIndices`
TL;DR: The per-byte-cost of `findIndices` (strict bytestring function) has a huge constant factor, rendering lineSplit inefficient

When looking at examples of various optimizations of Streams, I happened to notice a huge discrepancy between the performance of the same code using `lineSplit 1` versus `lines` (both from Data.ByteString.Streaming.Char8) when run over large input.

I checked the implementation source code for `lineSplit` and noticed that it calls the (strict) ByteString function `findIndices` to locate the *n*th newline character in a given byte sequence to capture everything up to and including that character. However, for `lineSplit 1` as a minimal example, this yields abysmal performance, as the per-byte cost of `findIndices` is ~20x greater than `findIndex`.

The call to `findIndices` should ideally be replaced with a function with lower per-byte overhead, which is a bug against the `bytestring` package. ([CR](https://github.com/haskell/bytestring/issues/269))

### PR DESCRIPTION
**Pull request by archaephyrryx**
findIndices vastly inefficient as currently implemented in bytestring
library

elemIndices is much faster for finding only newlines

should resolve #14

### CODE REVIEW

File: Data/ByteString/Streaming/Char8.hs [hunk 1]
```diff
@@ -654,7 +654,7 @@ lineSplit !n0 text0 = loop1 0 text0
           let !numNewlines = B.count newline c
               !newCounter = counter + numNewlines
            in if newCounter >= n
-                then case Prelude.drop (n - counter - 1) (B.findIndices (== newline) c) of
+                then case Prelude.drop (n - counter - 1) (B.elemIndices newline c) of
```
Comment (vdukhovni) [inline Data/ByteString/Streaming/Char8.hs]: It turns out that this is useful, but not nearly enough.  The current behaviour is quadratic in the chunk size after all, because 3 lines above this change are strictly counting the number of newlines in the chunk (it helps that `B.count` is more efficient than `B.findIndices`, but not by enough).

Therefore, when `n` (the requested number of lines to include in each batch, worst case 1) is small, but the 32k chunks contain many, say `m`, short lines, the runtime is O(m^2/n) after all.  The simplest solution appears to be to use `findIndex` in a new loop here, stopping when we either find no more new lines in the current chunk, or have found enough to reach the requested total, taking appropriate actions in each of the two cases.

Testing an approach along these lines I brought the `n = 1` runtime for a file with ~10 million  lines, averaging 126 bytes each, from 134s with the original code, and 120s with this PR, ultimately down to 7.7s after eliminating the `B.count` call, and instead looping `B.findIndex` for *up-to* the required number of times.

But even the _simplest_ approach above isn't the best one, because we're then we're "working too hard" when do want to find a batch of newlines for `n >> 1`.  What we really want is a function that will return either the a count of newlines (if smaller than the desired count) or else the total length of the initial substring holding the desired number of lines, to whit:
```haskell
-- | Return either how many newlines a strict bytestring chunk contains, if
-- fewer than the number requested, or, else the total length of the requested
-- number of lines within the bytestring (equivalently, the offset of the start
-- of the first /unwanted line/).
nthNewLine :: B.ByteString   -- input chunk
           -> Int            -- remaining number of newlines wanted
           -> Either Int Int -- Left count, else Right length
nthNewLine (B.PS fp off len) n =
  unsafeDupablePerformIO $ withForeignPtr fp $ \base ->
  loop (base `plusPtr` off) n 0 len
    where
      loop :: Ptr Word8 -> Int -> Int -> Int -> IO (Either Int Int)
      loop !_ 0 !o !_ = return $ Right o
      loop !p !i !o !l = do
        !q <- B.memchr p 10 $ fromIntegral l
        if q == nullPtr
            then return $! Left $! n - i
            else let !q' = q `plusPtr` 1
                     !d  = q' `minusPtr` p
                     !o' = o + d
                     !l' = l - d
                  in loop q' (i-1) o' l'
```
With this we not only get non-trivially faster performance for `n == 1` (~6.3s for the same input), but also better performance (2x) for e.g. `n == 400` where typically each chunk contains sufficiently many lines, but still often enough some final chunk only needs to deliver a smaller number of lines and we fall into the expensive `findIndices` code path in the tail chunk.

Furthermore, and somewhat surprisingly at first, I see better performance even with batches of 80 thousand lines, which is counter-intuitive, because one might expect the `B.count` performance to dominate the cost, and `B.count` should ideally be faster than repeated calls to `B.memchr` to do the same count, *but* internally `B.count` evaluates every byte comparing it with the desired value, while the C `memchr()` function is often an assembly optimised compiler intrinsic function that uses more efficient CPU instructions, and so outperforms `B.count` when the character you're looking for is not too frequent in the input.  The worst case performance should therefore be for input entirely consisting of newline characters (empty lines). 

Thus, it is only with 100 million empty lines that I see a small advantage in the `B.count` variant, which processes chunks of 80k lines in 0.6 seconds vs. 0.75s for the version that iterates `B.memchr`. 

So on balance, I think that the `B.memchr` version is the all-round performer, and we should not optimize for vast streams of empty lines, where the performance gain of the count variant is still modest.

Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs]: With `nthNewLine` in hand (feel free to assign a better name if one comes to mind), it is easy to substantially simplify `loop2`, and make it from 2x to 30x times over a broad range of line group sizes and input streams, with only a stream of just newlines losing out slightly.  I'm leaving that as an exercise.

Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs]: It might also be worth noting that as of `bytestring 0.11` the the `streaming-bytestring` package will no longer work with GHC 7.x, and so the cabal file needs a patch to cap the `bytestring` version to `< 0.11`.  Something along the lines of:
```
diff --git a/streaming-bytestring.cabal b/streaming-bytestring.cabal
index 435acd4..54bec8c 100644
--- a/streaming-bytestring.cabal
+++ b/streaming-bytestring.cabal
@@ -204,8 +204,10 @@ library
   if impl(ghc < 7.8)
     build-depends: bytestring < 0.10.4.0
                  , bytestring-builder
+  elif impl(ghc < 8.0)
+    build-depends: bytestring >= 0.10.4 && < 0.11
   else
-    build-depends: bytestring >= 0.10.4
+    build-depends: bytestring >= 0.10.4 && < 0.12
 
   if impl(ghc < 8.0)
     build-depends: semigroups
```

**Review by vdukhovni — CHANGES_REQUESTED** [review]

File: Data/ByteString/Streaming/Char8.hs (hunk 1 above)
Comment (archaephyrryx) [reply Data/ByteString/Streaming/Char8.hs]: I ended up reimplementing the function using most of the original logic, but optimizing out the repeated calls to `B.count` by creating a separate internal loop per chunk that "remembers" how many newlines there were originally when it was originally called, and decrements the value by the appropriate amount for each recursive call.

I haven't been able to obtain a practical performance metric of the improvement yet, but the new version compiles and passes the current test-suite on my local system, so it would be possible to compare the performance before and after.

**Comment by vdukhovni** [comment]
TL;DR:  THANKS! This PR is substantial progress, and never does worse and often does much better than the head master commit. I think that perhaps the next step should be taken, and the `B.count` code removed entirely, at least until it can be improved and shown to actually deliver speedups in realistic use cases.

I made some measurements of the resulting performance on the head commit of the master branch, this PR (15), and a version that (as proposed) entirely does away with calling `B.count` (the "No Count" columns below).  The measurements were made with `lineSplit` counts of 1, 256, 4096, 32768 and 524288 (16 x 32768) lines, and for a "normal" file with ~10m lines averaging 126 bytes and a worst-case file consisting of just empty lines.  The results are below:

Time to read and output to /dev/null 10million lines of ~126 byte lines, split
into chunks of 1, 256, 4096, 32768 and 524288 lines via `lineSplit`.

| SplitCount | Release      | PR#15   | No Count   |
| :--------- |  ----------: | ------: | ---------: |
| 1          | 131.331      | 4.450   | 3.630      |
| 256        | 9.480        | 1.651   | 0.771      |
| 4096       | 1.941        | 1.421   | 0.711      |
| 32768      | 1.481        | 1.390   | 0.700      |
| 524288     | 1.421        | 1.400   | 0.690      |

Time to read and output to /dev/null 10million empty lines, split into chunks
of 1, 256, 4096, 32768 and 524288 lines (i.e. bytes) via `lineSplit`.

| SplitCount | Release      | PR#15   | No Count   |
| :--------- |  ----------: | ------: | ---------: |
| 1          | 111.140      | 1.661   | 1.580      |
| 256        | 0.770        | 0.300   | 0.170      |
| 4096       | 0.320        | 0.280   | 0.160      |
| 32768      | 0.120        | 0.110   | 0.160      |
| 524288     | 0.100        | 0.101   | 0.160      |

No major surprises, the PR#15 code and original code do better on a worst-case stream of just empty lines, but only when the requested line cluster size is as big or bigger than the streaming bytestring chunk size, in which case they're able to take advantage of the call to `B.count` to save some work.

My instinct is that twice the performance on typical files with longer lines compares favourably against 1.6x the performance on a stream of empty lines seen only when the requested line grouping ~32k lines at a time or more!

For the record, here's the benchmark code:
```haskell
module Main (main) where

import qualified Streaming as S
import qualified Data.ByteString.Streaming.Char8 as Q
import qualified Data.List as L
import System.Environment (getArgs)

defaultLineCount :: Int
defaultLineCount = 1

main :: IO ()
main = do
    n <- maybe defaultLineCount (read . fst) <$> L.uncons <$> getArgs
    S.mapsM_ Q.putStr
        $ Q.lineSplit n
        $ Q.stdin
```

At this point feedback from the maintainers, et. al., would be great.

Cc: @bodigrim, @cartazio, @sjakobi, @andrewthad

**Comment by vdukhovni** [comment]
Anyone?

**Comment by andrewthad** [comment]
I'm not involved in the maintenance of this library anymore. That aside, I think that this seems like a good improvement. The test suite has a few tests for splitting on newlines, so if it's still passing, the new implementation is probably correct.

**Comment by vdukhovni** [comment]
> I'm not involved in the maintenance of this library anymore. That aside, I think that this seems like a good improvement. The test suite has a few tests for splitting on newlines, so if it's still passing, the new implementation is probably correct.

Who are the current (active?) maintainers?  It would be nice to see this resolved.  The key question is whether to go with this PR as-is, or to take a further step in the direction of eliminating the line count entirely, yielding broadly another ~2x performance improvement, at the cost of some loss of performance on vast streams of just newlines consumed with split counts larger than the number of lines (i.e. bytes) per chunk.

My take is that the "No Count" variant is the logical next step.

**Comment by vdukhovni** [comment]
Cc: @chessai, @fosskers any feedback on this PR? It seems the package has been stable and not in need of any maintenance for some time, which is often a feature (stable, well tested code, ...), but it would still be nice to get some dusty corners attended to...

**Comment by fosskers** [comment]
Hi @vdukhovni , sorry that it took a while to respond. I like where this PR is going, and think the `No Count` variant is the direction we should move in. As @andrewthad said, if the tests are passing and your benches show good numbers, then it should be an easy merge.

It can make a Hackage release as soon as all that is ready.

**Comment by vdukhovni** [comment]
> Hi @vdukhovni , sorry that it took a while to respond. I like where this PR is going, and think the `No Count` variant is the direction we should move in. As @andrewthad said, if the tests are passing and your benches show good numbers, then it should be an easy merge.
> 
> It can make a Hackage release as soon as all that is ready.

Done in #18 

I think the tests however could use some upkeep. They're using rather dated `stack` snapshots, and I think `cabal` would  be a better choice, and would ideally test `GHC 8.0, 8.2, 8.4, 8.6, 8.8 and 8.10` and soon also 9.0, as well as any `7.x` versions still supported, ... 
But otherwise, and perhaps with less effort a more comprehensive set of `stack` LTS snaphots and a more recent `nightly`.

**Comment by fosskers** [comment]
Oh it looks like I haven't given this repo the usual multi-LTS Github CI setup I use. I'll do that.

**Comment by archaephyrryx** [comment]
As #18 has already been merged, and this PR is a variant solution to the same problem addressed in that PR, I am closing both this PR and the associated issue (assuming the latter has not already been resolved).
